Do not assert page is evicted from back forward cache

szewai opened the pull request:
The test aims to verify that a page with open database connection can enter back forward cache, and new open database request from an active page does not get blocked by open connection of cached page. This is the behavior implemented in Chrome, Firefox and Safari. However, the test also asserts the cached page is removed from back forward cache. The eviction is unnecessary because request can be unblocked as long as the connection on cached page is closed. Currently Chrome and Firefox evict page from cache but Safari does not. Since this behavior is implementation-defined, we should drop the assertion from the test.

Also, update the tests to use different database names so they don't affect each other.

szewai commented:
@lozy219 @inexorabletash @asutherland What do you think?

inexorabletash commented:
I'm not going to be able to review this for several weeks. Maybe @fergald or @ayuishii could help

szewai committed:
Do not assert page is evicted from back forward cache

The test aims to verify that a page with open database connection can enter back
forward cache, and new open database request from an active page does not get
blocked by open connection of cached page. This is the behavior implemented in
Chrome, Firefox and Safari. However, the test also asserts the cached page is
removed from back forward cache. The eviction is unnecessary because request can
be unblocked as long as the connection on cached page is closed. Currently
Chrome and Firefox evict page from cache but Safari does not. Since this
behavior is implementation-defined, we should drop the assertion from the test.

Also, update the tests to use different database names so they don't affect each
other.

szewai commented:
> I'm not going to be able to review this for several weeks. Maybe @fergald or @ayuishii could help

I landed the fix to solve the confusing failure on WebKit's end. Please let me know if you get time to review and spot an issue.

fergald commented:
I don't think it's OK to allow the page to stay in cache after a version change. If the page is restored, it will attempt to run its `onversionchange` handler *after* the upgrade has happened. This results in a correctness problem, e.g. the [MDN example](https://w3c.github.io/IndexedDB/#handling-versionchange) would perform a write to the DB using the old schema and possibly corrupt the DB.

BFCache behaviour is not spelled out in the IndexedDB spec and the test should have better comments but correctness must come before caching.

If I am not mistaken about the behaviour, I think this change should be reverted.

szewai commented:
> I don't think it's OK to allow the page to stay in cache after a version change. If the page is restored, it will attempt to run its `onversionchange` handler _after_ the upgrade has happened. This results in a correctness problem, e.g. the [MDN example](https://w3c.github.io/IndexedDB/#handling-versionchange) would perform a write to the DB using the old schema and possibly corrupt the DB.
> 
> BFCache behaviour is not spelled out in the IndexedDB spec and the test should have better comments but correctness must come before caching.
> 
> If I am not mistaken about the behaviour, I think this change should be reverted.

Safari / WebKit does not fire version change event; it fires error event on pending requests and fires close event (this is similar to what we do when user deletes website data or process crash). Web page could reopen database connection on close.

If there is a behavior we want to ensure, we should include it in test. I am not opposed to evicting page from cache, if that's the only way we could achieve the behavior; but I think we should test behavior (like version change event handler is not invoked when going back), not implementation (like asserting page is evicted from cache).

cdumez commented:
> I am not opposed to evicting page from cache
Actually we are opposed to it. It is not acceptable to reduce the efficiency of WebKit's back/forward cache.

fergald commented:
> If there is a behavior we want to ensure, we should include it in test.

The test did test this, the `onversionchange` handler is set to cause a failure if it fires but the test no longer navigates back., so this is not tested with the change from `assertNotRestoredFromBFCache` to `waitUntilIndexedDBOpenForTesting`. I think restoring a back navigation to the test would address that on all browsers.

Firing `error` seems fine from a strict correctness perspective. What error type do you use? Does any page correctly handle this beyond telling the user that something went wrong and they should reload the page?

szewai commented:
> > If there is a behavior we want to ensure, we should include it in test.
> 
> The test did test this, the `onversionchange` handler is set to cause a failure if it fires but the test no longer navigates back., so this is not tested with the change from `assertNotRestoredFromBFCache` to `waitUntilIndexedDBOpenForTesting`. I think restoring a back navigation to the test would address that on all browsers.

Oh I see: `createIndexedDBForTesting` sets `onversionchange` set on the `IDBRequest`, which intends to ensure versionchange event is not fired. Well the behavior is actually not tested before my change because versionchange event is dispatched to `IDBDatabase`, not `IDBRequest`. I will update the test to cover that.

> 
> Firing `error` seems fine from a strict correctness perspective. What error type do you use? Does any page correctly handle this beyond telling the user that something went wrong and they should reload the page?

We currently use UnknownError. I don't know any real-world use case for this, but it seems the web page's choice if they prefer reloading / promoting user on IndexedDB error / close event instead of retrying operations silently.

szewai commented:
> > > If there is a behavior we want to ensure, we should include it in test.
> > 
> > 
> > The test did test this, the `onversionchange` handler is set to cause a failure if it fires but the test no longer navigates back., so this is not tested with the change from `assertNotRestoredFromBFCache` to `waitUntilIndexedDBOpenForTesting`. I think restoring a back navigation to the test would address that on all browsers.
> 
> Oh I see: `createIndexedDBForTesting` sets `onversionchange` set on the `IDBRequest`, which intends to ensure versionchange event is not fired. Well the behavior is actually not tested before my change because versionchange event is dispatched to `IDBDatabase`, not `IDBRequest`. I will update the test to cover that.
> 

Filed https://github.com/web-platform-tests/wpt/pull/47699.
